debian/bsondump.1
.\" Man page generated from reStructuredText.
.
.TH "BSONDUMP" "1" "Oct 24, 2019" "4.0" "mongodb-manual"
.SH NAME
bsondump \- MongoDB BSON Utility
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SS On this page
.INDENT 0.0
.IP \(bu 2
\fI\%Synopsis\fP
.IP \(bu 2
\fI\%Options\fP
.IP \(bu 2
\fI\%Examples\fP
.UNINDENT
.INDENT 0.0
.INDENT 3.5
.IP "macOS Sierra and Go 1.6 Incompatibility"
.sp
Users running on macOS Sierra require the 3.2.10 or newer version
of  \fI\%bsondump\fP\&.
.UNINDENT
.UNINDENT
.SH SYNOPSIS
.sp
The \fI\%bsondump\fP converts BSON files into human\-readable
formats, including JSON\&. For example, \fI\%bsondump\fP is useful
for reading the output files generated by \fBmongodump\fP\&.
.sp
Run \fI\%bsondump\fP from the system command line, not the \fBmongo\fP shell.
.sp
\fBIMPORTANT:\fP
.INDENT 0.0
.INDENT 3.5
\fI\%bsondump\fP is a diagnostic tool for inspecting
BSON files, not a tool for data ingestion or other application use.
.UNINDENT
.UNINDENT
.SH OPTIONS
.sp
Changed in version 3.0.0: \fI\%bsondump\fP removed the \fB\-\-filter\fP, \fB\-\-dbpath\fP and the
\fB\-\-noobjcheck\fP options.

.INDENT 0.0
.TP
.B bsondump
.UNINDENT
.INDENT 0.0
.TP
.B \-\-help
Returns information on the options and use of \fBbsondump\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-verbose, \-v
Increases the amount of internal reporting returned on standard output
or in log files. Increase the verbosity with the \fB\-v\fP form by
including the option multiple times, (e.g. \fB\-vvvvv\fP\&.)
.UNINDENT
.INDENT 0.0
.TP
.B \-\-quiet
Runs \fBbsondump\fP in a quiet mode that attempts to limit the amount
of output.
.sp
This option suppresses:
.INDENT 7.0
.IP \(bu 2
output from database commands
.IP \(bu 2
replication activity
.IP \(bu 2
connection accepted events
.IP \(bu 2
connection closed events
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \-\-version
Returns the \fBbsondump\fP release number.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-objcheck
Validates each BSON object before outputting it in JSON
format. By default, \fBbsondump\fP enables \fI\%\-\-objcheck\fP\&.
For objects with a high degree of sub\-document nesting,
\fI\%\-\-objcheck\fP can have a small impact on performance.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-type <=json|=debug>
Changes the operation of \fBbsondump\fP from outputting
"JSON" (the default) to a debugging format.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-pretty
New in version 3.0.0.

.sp
Outputs documents in a pretty\-printed format JSON.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-bsonFile
New in version 3.4.

.sp
Specifies a path to a BSON file to dump to JSON. \fI\%\-\-bsonFile\fP is
an alternative to the positional \fI\%<bsonFilename>\fP option.
.sp
By default, \fBbsondump\fP reads from standard input.
.UNINDENT
.INDENT 0.0
.TP
.B <bsonFilename>
The final argument to \fBbsondump\fP is a document containing
BSON\&. This data is typically generated by
\fBbsondump\fP or by MongoDB in a rollback operation.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-outFile
New in version 3.4.

.sp
Specifies the path of the file to which \fBbsondump\fP should write
its output JSON data.
.sp
By default, \fBbsondump\fP writes to standard output.
.UNINDENT
.SH EXAMPLES
.sp
Changed in version 3.4.

.sp
By default, \fI\%bsondump\fP outputs data to standard output. To
create corresponding JSON files, you can use the
\fI\%\-\-outFile\fP option:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
bsondump \-\-outFile collection.json collection.bson
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Use the following command (at the system shell) to produce debugging
output for a BSON file:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
bsondump \-\-type=debug collection.bson
.ft P
.fi
.UNINDENT
.UNINDENT
.SH AUTHOR
MongoDB Documentation Project
.SH COPYRIGHT
2008-2019
.\" Generated by docutils manpage writer.
.


NEWS
Ticket reference tags refer to tickets in the MongoDB JIRA tracking system:
https://jira.mongodb.org

WiredTiger release 2.9.3, 2017-06-27
------------------------------------

See the upgrading documentation for details of API and behavior changes.

Significant changes:
* WT-2972       Add an interface allowing partial updates to existing values
* WT-3063       Add an interface allowing reservation of records for read-modify-write
* WT-3142       Add a workload generator application
* WT-3160       Improve eviction of internal pages from idle trees
* WT-3245       Avoid hangs on shutdown when a utility thread encounters an error
* WT-3258       Improve visibility into thread wait time due to pages exceeding memory_page_max
* WT-3263       Allow archive on restart/recovery if clean shutdown
* WT-3287       Review WiredTiger internal panic checks
* WT-3292       Review/cleanup full-barrier calls in WiredTiger
* WT-3296       LAS table fixes/improvements
* WT-3327       Checkpoints can hang if time runs backward
* WT-3345       Improve rwlock scaling
* WT-3373       Access violation due to a bug in internal page splitting
* WT-3379       Change when pages can be split to avoid excessively slowing some operations

See JIRA changelog for a full listing:
https://jira.mongodb.org/browse/WT/fixforversion/18291

WiredTiger release 2.9.2, 2017-05-25
------------------------------------

See the upgrading documentation for details of API and behavior changes.

Significant changes:
* SERVER-16796  Increase logging activity for journal recovery operations
* SERVER-28168  Cannot start or repair MongoDB after unexpected shutdown.
* SERVER-28194  Fix a bug where a missing WiredTiger.turtle file can result in data loss
* WT-98         Update the current cursor value without a search
* WT-2439       Enhance reconciliation page layout
* WT-2898       Improve performance of eviction-heavy workloads by dynamically controlling the number of eviction threads
* WT-3097       Race on reconfigure or shutdown can lead to waiting for statistics log server
* WT-3106       Add truncate support to command line wt utility
* WT-3114       Avoid archiving log files immediately after recovery
* WT-3115       Change the dhandle lock to a read/write lock
* WT-3127       Fix a bug with CPU yield calls don't necessarily imply memory barriers
* WT-3135       Fix a bug with search_near() for index with custom collator
* WT-3137       Fix a hang in __log_slot_join/__log_slot_switch_internal
* WT-3144       Fix a bug where random cursor returns not-found when descending to an empty page
* WT-3148       Improve eviction efficiency with many small trees
* WT-3149       Change eviction to start new walks from a random place in the tree
* WT-3150       Reduce impact of checkpoints on eviction server
* WT-3152       Convert table lock from a spinlock to a read write lock
* WT-3157       Fix a bug in checkpoint/transaction integrity issue when writes fail.
* WT-3158       Fix structure packing on Windows.
* WT-3188       Fix error handling in logging where fatal errors could lead to a hang
* WT-3193       Close a race between verify opening a handle and eviction visiting it
* WT-3206       Fix a bug where core dump happened on NULL page index
* WT-3218       Fix a bug that could lead to unexpected checkpoint ordering failures
* WT-3243       Reorder log slot release so joins don't wait on IO
* WT-3262       Stop making schema operations wait for cache

See JIRA changelog for a full listing:
https://jira.mongodb.org/browse/WT/fixforversion/17874


WiredTiger release 2.9.1, 2016-12-22
------------------------------------

New features and API changes; refer to the API documentation for full details:

* SERVER-26545	Remove fixed-size limitation on WiredTiger hazard pointers. See the upgrading documentation for details
* WT-283	Add a new WT_SESSION::alter method that can be used to reconfigure table metadata
* WT-2670	Change the default file system access pattern advice for data files from random to no advice. Add access_pattern_hint configuration option for WT_SESSION::create API that can be used to advise the file system of expected access semantics. See the upgrading documentation for details.
* WT-3034	Add support for including updates when reading from named snapshots

Significant changes and bug fixes:

* WT-2960	Reduce likelihood of using the lookaside file, especially when inserting multi-megabyte values
* WT-3056	Allow projected table and join cursors to use primary keys
* WT-3070	Fix a bug in search_near on indexes

Other noteworthy changes since the previous release:

* WT-2336	Add a test validating schema operations via file system call monitoring
* WT-2402	Pad structures to avoid cache line sharing
* WT-2771	Add a statistic to track per-btree dirty cache usage
* WT-2833	Add projections to wt dump utility
* WT-2969	Possible snapshot corruption during compaction
* WT-3014	Add GCC/clang support for ELF symbol visibility
* WT-3021	Fixes for java log example, raw mode in java, and raw mode in log cursors
* WT-3025	Fix error path in log_force_sync
* WT-3028	Don't check for blocked eviction with in-memory workloads
* WT-3030	Fix a race between scans and splits reading the index hint
* WT-3037	Clean up some log slot comments
* WT-3048	WiredTiger maximum size warning uses the wrong format
* WT-3051	Remove external __wt_hex symbol
* WT-3052	Improve search if index hint is wrong
* WT-3053	Make Python use internal memory allocation again
* WT-3054	Make a PackOutputStream constructor that is compatible with the previous interface.
* WT-3055	When an AsyncOp is created, cache the whether the cursor is "raw"
* WT-3057	WiredTiger hazard pointers should use the WT_REF, not the WT_PAGE
* WT-3061	Syscall testing should support pwrite64 on Linux
* WT-3064	Minor tree cleanups: .gitignore, NEWS misspelling
* WT-3066	Minor code cleanups
* WT-3068	Copy artifacts of test runs in wtperf_run script

tools/bsondump.cpp
// restore.cpp

/**
*    Copyright (C) 2008 10gen Inc.
*
*    This program is free software: you can redistribute it and/or  modify
*    it under the terms of the GNU Affero General Public License, version 3,
*    as published by the Free Software Foundation.
*
*    This program is distributed in the hope that it will be useful,
*    but WITHOUT ANY WARRANTY; without even the implied warranty of
*    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*    GNU Affero General Public License for more details.
*
*    You should have received a copy of the GNU Affero General Public License
*    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "../pch.h"
#include "../client/dbclient.h"
#include "../util/mmap.h"
#include "../util/text.h"
#include "tool.h"

#include <boost/program_options.hpp>

#include <fcntl.h>

using namespace mongo;

namespace po = boost::program_options;

class BSONDump : public BSONTool {

    enum OutputType { JSON , DEBUG } _type;

public:

    BSONDump() : BSONTool( "bsondump", NONE ) {
        add_options()
        ("type" , po::value<string>()->default_value("json") , "type of output: json,debug" )
        ;
        add_hidden_options()
        ("file" , po::value<string>() , ".bson file" )
        ;
        addPositionArg( "file" , 1 );
        _noconnection = true;
    }

    virtual void printExtraHelp(ostream& out) {
        out << "Display BSON objects in a data file.\n" << endl;
        out << "usage: " << _name << " [options] <bson filename>" << endl;
    }

    virtual int doRun() {
        {
            string t = getParam( "type" );
            if ( t == "json" )
                _type = JSON;
            else if ( t == "debug" )
                _type = DEBUG;
            else {
                cerr << "bad type: " << t << endl;
                return 1;
            }
        }

        path root = getParam( "file" );
        if ( root == "" ) {
            printExtraHelp(cout);
            return 1;
        }

        processFile( root );
        return 0;
    }

    bool debug( const BSONObj& o , int depth=0) {
        string prefix = "";
        for ( int i=0; i<depth; i++ ) {
            prefix += "\t\t\t";
        }

        int read = 4;

        try {
            cout << prefix << "--- new object ---\n";
            cout << prefix << "\t size : " << o.objsize() << "\n";
            BSONObjIterator i(o);
            while ( i.more() ) {
                BSONElement e = i.next();
                cout << prefix << "\t\t " << e.fieldName() << "\n" << prefix << "\t\t\t type:" << setw(3) << e.type() << " size: " << e.size() << endl;
                if ( ( read + e.size() ) > o.objsize() ) {
                    cout << prefix << " SIZE DOES NOT WORK" << endl;
                    return false;
                }
                read += e.size();
                try {
                    e.validate();
                    if ( e.isABSONObj() ) {
                        if ( ! debug( e.Obj() , depth + 1 ) )
                            return false;
                    }
                    else if ( e.type() == String && ! isValidUTF8( e.valuestr() ) ) {
                        cout << prefix << "\t\t\t" << "bad utf8 String!" << endl;
                    }
                    else if ( logLevel > 0 ) {
                        cout << prefix << "\t\t\t" << e << endl;
                    }

                }
                catch ( std::exception& e ) {
                    cout << prefix << "\t\t\t bad value: " << e.what() << endl;
                }
            }
        }
        catch ( std::exception& e ) {
            cout << prefix << "\t" << e.what() << endl;
        }
        return true;
    }

    virtual void gotObject( const BSONObj& o ) {
        switch ( _type ) {
        case JSON:
            cout << o.jsonString( TenGen ) << endl;
            break;
        case DEBUG:
            debug(o);
            break;
        default:
            cerr << "bad type? : " << _type << endl;
        }
    }
};

int main( int argc , char ** argv ) {
    BSONDump dump;
    return dump.main( argc , argv );
}
